interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/BaseNodeCompiler.scala

```diff
@@ -0,0 +1,325 @@
+
+    combineErrors(nodeCompilation, additionalValidationResult)
+  }
+
+  def compileSwitch(
+      expressionRaw: Option[(String, Expression)],
+      choices: List[(String, Expression)],
+      ctx: ValidationContext
+  )(
+      implicit nodeId: NodeId
+  ): NodeCompilationResult[(Option[expression.Expression], List[expression.Expression])] = {
+
+    // the frontend uses empty string to delete deprecated expression.
+    val expression = expressionRaw.filterNot(_._1.isEmpty)
+
+    val expressionCompilation = expression.map { case (output, expression) =>
+      ExpressionCompilerAdapter
+        .compileExpression(
+          expression,
+          ctx,
+          Unknown,
+          NodeExpressionId.DefaultExpressionId,
+          Some(OutputVar.switch(output))
+        )
+        ._2
+    }
+    val objExpression = expressionCompilation.map(_.compiledObject.map(Some(_))).getOrElse(Valid(None))
+
+    val caseCtx = expressionCompilation.flatMap(_.validationContext.toOption).getOrElse(ctx)
+
+    val (additionalValidations, caseExpressions) = choices.map { case (outEdge, caseExpr) =>
+      val (expressionCompilation, nodeCompilation) =
+        ExpressionCompilerAdapter.compileExpression(caseExpr, caseCtx, Typed[Boolean], outEdge, None)
+      val typedExpression = expressionCompilation.typedExpression
+      val validation      = ValidationAdapter.validateMaybeBoolean(typedExpression, outEdge)
+      val caseExpression  = nodeCompilation
+      (validation, caseExpression)
+    }.unzip
+
+    val expressionTypingInfos = caseExpressions
+      .map(_.expressionTypingInfo)
+      .foldLeft(expressionCompilation.map(_.expressionTypingInfo).getOrElse(Map.empty)) {
+        _ ++ _
+      }
+
+    val objCases = caseExpressions.map(_.compiledObject).sequence
+
+    val compilationResult = NodeCompilationResult(
+      expressionTypingInfos,
+      None,
+      expressionCompilation.map(_.validationContext).getOrElse(Valid(ctx)),
+      objExpression.product(objCases),
+      expressionCompilation.flatMap(_.expressionType)
+    )
+
+    combineErrors(compilationResult, additionalValidations.combineAll)
+  }
+
+  def compileFields(
+      fields: List[pl.touk.nussknacker.engine.graph.variable.Field],
+      ctx: ValidationContext,
+      outputVar: Option[OutputVar]
+  )(implicit nodeId: NodeId): NodeCompilationResult[List[variable.Field]] = {
+
+    final case class CompiledRecordField(
+        field: pl.touk.nussknacker.engine.graph.variable.Field,
+        index: Int,
+        typedExpression: TypedExpression
+    )
+
+    val compliedRecordFields = fields.zipWithIndex.map { case (field, index) =>
+      objectParametersExpressionCompiler
+        .compile(field.expression, Some(node.recordValueFieldName(index)), ctx, Unknown)
+        .map(CompiledRecordField(field, index, _))
+    }
+
+    val recordValuesCompilationResult = compliedRecordFields.traverse { validatedField =>
+      validatedField.andThen { compiledField =>
```

Comment:
It can be just `.map {` (without wrapping with `Valid`)

Reply:
done